icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -0,0 +1,251 @@
+            snFirstPatterns = PersonNamePattern.makePatterns(asStringArray(snFirstResource), this);
+
+            gnFirstLocales = new HashSet<>();
+            Collections.addAll(gnFirstLocales, asStringArray(rb.getWithFallback("personNames/nameOrderLocales/givenFirst")));
+            snFirstLocales = new HashSet<>();
+            Collections.addAll(snFirstLocales, asStringArray(rb.getWithFallback("personNames/nameOrderLocales/surnameFirst")));
+        } else {
+            ICUResourceBundle patternResource = rb.getWithFallback(RESOURCE_PATH_PREFIX + "sorting-" + resourceNameBody);
+
+            gnFirstPatterns = PersonNamePattern.makePatterns(asStringArray(patternResource), this);
+            snFirstPatterns = null;
+            gnFirstLocales = null;
+            snFirstLocales = null;
+        }
+    }
+
+    public String format(PersonNameFormatter.PersonName name) {
+        // TODO: Should probably return a FormattedPersonName object
+
+        // if the formatter is for a language that doesn't use spaces between words and the name is from a language
+        // that does, create a formatter for the NAME'S locale and use THAT to format the name
+        ULocale nameLocale = name.getNameLocale();
+        boolean nameLocaleUsesSpaces = !LOCALES_THAT_DONT_USE_SPACES.contains(nameLocale.getLanguage());
+        if (!formatterLocaleUsesSpaces && nameLocaleUsesSpaces) {
+            PersonNameFormatterImpl nativeFormatter = new PersonNameFormatterImpl(nameLocale, this.length,
+                    this.usage, this.formality, this.options);
+            String result = nativeFormatter.format(name);
+
+            // BUT, if the name is actually written in the formatter locale's script, replace any spaces in the name
+            // with the foreignSpaceReplacement character
+            if (!foreignSpaceReplacement.equals(" ") && scriptMatchesLocale(result, this.locale)) {
+                result = result.replace(" ", this.foreignSpaceReplacement);
+            }
+            return result;
+        }
+
+        // if we get down to here, we're just doing normal formatting-- if we have both GN-first and SN-first rules,
+        // choose which one to use based on the name's locale and preferred field order
+        if (snFirstPatterns == null || nameIsGnFirst(name)) {
+            return getBestPattern(gnFirstPatterns, name).format(name);
+        } else {
+            return getBestPattern(snFirstPatterns, name).format(name);
+        }
+    }
+
+    public ULocale getLocale() {
+        return locale;
+    }
+
+    public String getInitialPattern() {
+        return initialPattern;
+    }
+
+    public String getInitialSequencePattern() {
+        return initialSequencePattern;
+    }
+
+    public boolean shouldCapitalizeSurname() {
+        return capitalizeSurname;
+    }
+
+    private final Set<String> LOCALES_THAT_DONT_USE_SPACES = new HashSet<>(Arrays.asList("ja", "zh", "th", "yue"));
```

Comment:
Do you want to hard-code this? There might be other languages, e.g. `km` that don't use spaces. Perhaps it can be derived?

Reply:
I added `km` and `lo`.  I don't like hard-coding this, but as far as I know, we don't have any existing resources anywhere else that'll tell us what we need to know here, and I'm not sure how we'd derive it algorithmically.  If you guys have ideas, I'd love to hear them.